File: mobile/src/db/schema.ts
export const MIGRATIONS = [
  // v1 - initial schema
  `CREATE TABLE IF NOT EXISTS notes (
    id TEXT PRIMARY KEY NOT NULL,
    user_id TEXT,
    folder_id TEXT,
    title TEXT NOT NULL DEFAULT 'Untitled',
    content TEXT NOT NULL DEFAULT '',
    is_pinned INTEGER NOT NULL DEFAULT 0,
    created_at TEXT,
    updated_at TEXT,
    deleted_at TEXT
  )`,
  `CREATE INDEX IF NOT EXISTS idx_notes_folder ON notes(folder_id)`,
  `CREATE INDEX IF NOT EXISTS idx_notes_updated ON notes(updated_at DESC)`,
  `CREATE INDEX IF NOT EXISTS idx_notes_pinned_updated ON notes(is_pinned DESC, updated_at DESC)`,
  `CREATE INDEX IF NOT EXISTS idx_notes_deleted ON notes(deleted_at)`,

  `CREATE TABLE IF NOT EXISTS folders (
    id TEXT PRIMARY KEY NOT NULL,
    user_id TEXT,
    name TEXT NOT NULL,
    created_at TEXT,
    updated_at TEXT
  )`,
  `CREATE INDEX IF NOT EXISTS idx_folders_updated ON folders(updated_at DESC)`,

  `CREATE TABLE IF NOT EXISTS tasks (
    id TEXT PRIMARY KEY NOT NULL,
    user_id TEXT,
    title TEXT NOT NULL DEFAULT 'Untitled',
    description TEXT NOT NULL DEFAULT '',
    is_completed INTEGER NOT NULL DEFAULT 0,
    due_date TEXT,
    created_at TEXT,
    updated_at TEXT,
    deleted_at TEXT
  )`,
  `CREATE INDEX IF NOT EXISTS idx_tasks_updated ON tasks(updated_at DESC)`,

  // v2 - sync_state table
  `CREATE TABLE IF NOT EXISTS sync_state (
    id TEXT PRIMARY KEY NOT NULL,
    user_id TEXT,
    device_id TEXT NOT NULL UNIQUE,
    last_cursor TEXT,
    last_sync_at TEXT,
    updated_at TEXT NOT NULL
  )`,
];

File: note-desktop/Sources/Utils/Schema.swift
import Foundation

public struct Schema {
    public static let createUsersTable = """
    CREATE TABLE IF NOT EXISTS users (
        id TEXT PRIMARY KEY,
        username TEXT,
        email TEXT UNIQUE
    );
    """
    
    public static let createFoldersTable = """
    CREATE TABLE IF NOT EXISTS folders (
        id TEXT PRIMARY KEY,
        name TEXT,
        userId TEXT,
        deletedAt TEXT
    );
    """
    
    public static let createNotesTable = """
    CREATE TABLE IF NOT EXISTS notes (
        id TEXT PRIMARY KEY,
        title TEXT,
        content TEXT,
        createdAt TEXT,
        updatedAt TEXT,
        isPinned INTEGER DEFAULT 0,
        folderId TEXT,
        userId TEXT,
        deletedAt TEXT
    );
    """
    
    public static let createSyncOpsTable = """
    CREATE TABLE IF NOT EXISTS sync_ops (
        id TEXT PRIMARY KEY,
        opType TEXT,
        entityType TEXT,
        entityId TEXT,
        payload TEXT,
        updatedAt TEXT
    );
    """
    
    public static let createNotesFtsTable = """
    CREATE VIRTUAL TABLE IF NOT EXISTS notes_fts USING fts5(id UNINDEXED, title, content);
    """
    
    public static let createNotesFtsInsertTrigger = """
    CREATE TRIGGER IF NOT EXISTS notes_ai AFTER INSERT ON notes BEGIN
        INSERT INTO notes_fts (id, title, content) VALUES (new.id, new.title, new.content);
    END;
    """
    
    public static let createNotesFtsUpdateTrigger = """
    CREATE TRIGGER IF NOT EXISTS notes_au AFTER UPDATE ON notes BEGIN
        UPDATE notes_fts SET title = new.title, content = new.content WHERE id = old.id;
    END;
    """
    
    public static let createNotesFtsDeleteTrigger = """
    CREATE TRIGGER IF NOT EXISTS notes_ad AFTER DELETE ON notes BEGIN
        DELETE FROM notes_fts WHERE id = old.id;
    END;
    """
    
    public static let allCreations = [
        createUsersTable,
        createFoldersTable,
        createNotesTable,
        createSyncOpsTable,
        createNotesFtsTable,
        createNotesFtsInsertTrigger,
        createNotesFtsUpdateTrigger,
        createNotesFtsDeleteTrigger
    ]
}
